libs/api/ebpf_api.cpp

```diff
@@ -1641,6 +1625,51 @@ _initialize_ebpf_object_from_elf(
     return result;
 }
 
+// Find a map that needs to be created and doesn't depend on
+// creating another map first.  That is, we want to create an
+// inner map template before creating an outer map that depends
+// on the inner map template.
+static ebpf_map_t*
+_get_next_map_to_create(std::vector<ebpf_map_t*>& maps)
+{
+    for (auto& map : maps) {
+        if (map->map_handle != ebpf_handle_invalid) {
+            // Already created.
+            continue;
+        }
+        if (map->map_definition.type != BPF_MAP_TYPE_ARRAY_OF_MAPS &&
+            map->map_definition.type != BPF_MAP_TYPE_HASH_OF_MAPS) {
+            return map;
+        }
+        if (map->inner_map == nullptr) {
+            // This map requires an inner map template, look up which one.
+            for (auto& inner_map : maps) {
+                if (inner_map->original_fd == map->inner_map_original_fd) {
+                    map->inner_map = inner_map;
+                    break;
+                }
+            }
+            if (map->inner_map == nullptr) {
+                // We can't create this map because there is no inner template.
```

Comment:
If inner map template is not present, then the whole program load should fail?

Reply:
I don't think it fails, I think it just doesn't create the outer map.

Reply:
I think the behavior should be all or none. libbpf also has similar behavior. If inner map template is not found, then we should return error.

Reply:
The same is true for programs also. If loading of one program fails, the whole API fails.

Reply:
Updated